pkg/typesystem/weighted_graph.go

```diff
@@ -0,0 +1,137 @@
+package typesystem
+
+import (
+	"fmt"
+	"slices"
+
+	"github.com/openfga/language/pkg/go/graph"
+
+	"github.com/openfga/openfga/internal/utils"
+)
+
+type weightedGraphItem interface {
+	GetWeight(destinationType string) (int, bool)
+}
+
+// hasPathTo returns a boolean indicating if a path exists from a node or edge to a terminal type. E.g
+// can we reach "user" from "document".
+func hasPathTo(nodeOrEdge weightedGraphItem, destinationType string) bool {
+	_, ok := nodeOrEdge.GetWeight(destinationType)
+	return ok
+}
+
+type IntersectionEdges struct {
+	LowestEdge *graph.WeightedAuthorizationModelEdge   // nil if direct is lowest, otherwise lowest edge
+	Siblings   []*graph.WeightedAuthorizationModelEdge // all non lowest and excluding direct edges siblings
+	// We need to separate out the direct edges from the other edges so that in the case of direct edges are not the lowest weight,
+	// the check against these direct edges do not require checking against these individual directly assigned edges.
+	// Instead, these directly assigned edges should be treated as a group such that if one of these edges satisfy,
+	// the intersection satisfy.
+	DirectEdges               []*graph.WeightedAuthorizationModelEdge
+	DirectEdgesAreLeastWeight bool // whether direct edges are the lowest weight edges
+}
+
+// GetEdgesForIntersection returns the lowest weighted edge and
+// its siblings edges for intersection based via the weighted graph.
+// If the direct edges have equal weight as its sibling edges, it will choose
+// the direct edges as preference.
+// If any of the children are not connected, it will return empty IntersectionEdges.
+func GetEdgesForIntersection(edges []*graph.WeightedAuthorizationModelEdge, sourceType string) (IntersectionEdges, error) {
+	if len(edges) < 2 {
+		// Intersection by definition must have at least 2 children
+		return IntersectionEdges{}, fmt.Errorf("invalid edges for source type %s", sourceType)
+	}
+
+	directEdges := make([]*graph.WeightedAuthorizationModelEdge, 0, len(edges))
+	var lowestEdge *graph.WeightedAuthorizationModelEdge
+	directEdgesAreLowest := false
+	lowestWeight := 0
+
+	// Process all edges first to separate direct from non-direct
+	nonDirectEdges := make([]*graph.WeightedAuthorizationModelEdge, 0, len(edges))
+
+	// It is assumed that direct edge always appear first.
+	for _, edge := range edges {
+		if !hasPathTo(edge, sourceType) {
+			continue
+		}
+		if edge.GetEdgeType() == graph.DirectEdge {
```

Comment:
```suggestion
		if edge.GetEdgeType() != graph.DirectEdge {
		  nonDirectEdges = append(nonDirectEdges, edge)
		  continue
		}
```

Then handle the non direct case.